packages/universal-test-runner/tests/run.test.ts

```diff
@@ -23,7 +30,143 @@ describe('Run function', () => {
+    let mockAdapter: Adapter
+
+    beforeEach(() => {
+      mockAdapter = { executeTests: jest.fn(() => ({ exitCode: 0 })) }
+    })
+
+    it('reads a list of test names', async () => {
+      await run(mockAdapter, { version: '0.1.0', testsToRun: 'test1|test4|test9' })
+
+      expect(mockAdapter.executeTests).toHaveBeenCalledWith({
+        testsToRun: [{ testName: 'test1' }, { testName: 'test4' }, { testName: 'test9' }],
+      })
+    })
+
+    it('reads the test suite for a single test', async () => {
+      await run(mockAdapter, { version: '0.1.0', testsToRun: 'test1|test4|suitename#test9' })
+
+      expect(mockAdapter.executeTests).toHaveBeenCalledWith({
+        testsToRun: [
+          { testName: 'test1' },
+          { testName: 'test4' },
+          { testName: 'test9', suiteName: 'suitename' },
+        ],
+      })
+    })
+
+    it('reads the test suite for a many tests', async () => {
+      await run(mockAdapter, {
+        version: '0.1.0',
+        testsToRun: 'suite1#test1|suite2#test4|suitename#test9',
+      })
+
+      expect(mockAdapter.executeTests).toHaveBeenCalledWith({
+        testsToRun: [
+          { testName: 'test1', suiteName: 'suite1' },
+          { testName: 'test4', suiteName: 'suite2' },
+          { testName: 'test9', suiteName: 'suitename' },
+        ],
+      })
+    })
+
+    it('reads the filepath only for many tests', async () => {
+      await run(mockAdapter, {
+        version: '0.1.0',
+        testsToRun: 'file1.js##test1|file2.js##test4|file3.js##test9',
+      })
+
+      expect(mockAdapter.executeTests).toHaveBeenCalledWith({
+        testsToRun: [
+          { testName: 'test1', filepath: 'file1.js' },
+          { testName: 'test4', filepath: 'file2.js' },
+          { testName: 'test9', filepath: 'file3.js' },
+        ],
+      })
+    })
+
+    it('reads the filepath only for one test', async () => {
+      await run(mockAdapter, { version: '0.1.0', testsToRun: 'test1|test4|file3.js##test9' })
+
+      expect(mockAdapter.executeTests).toHaveBeenCalledWith({
+        testsToRun: [
+          { testName: 'test1' },
+          { testName: 'test4' },
+          { testName: 'test9', filepath: 'file3.js' },
+        ],
+      })
+    })
+
+    it('reads the filepath and suite name for one test', async () => {
+      await run(mockAdapter, { version: '0.1.0', testsToRun: 'test1|test4|file3.js#suite1#test9' })
+
+      expect(mockAdapter.executeTests).toHaveBeenCalledWith({
+        testsToRun: [
+          { testName: 'test1' },
+          { testName: 'test4' },
+          { testName: 'test9', suiteName: 'suite1', filepath: 'file3.js' },
+        ],
+      })
+    })
+
+    it('reads an empty list when value is not present', async () => {
+      await run(mockAdapter, { version: '0.1.0' })
+
+      expect(mockAdapter.executeTests).toHaveBeenCalledWith({ testsToRun: [] })
+    })
+  })
+
+  it('reads test cases to run from a file', async () => {
+    vol.fromJSON(
```

Comment:
This is a much handier way to do fs mocking, will follow up in another PR to replace our existing ad-hoc fs mocks with this approach.